p2p/src/main/java/com/google/android/fhir/p2p/SyncPayload.kt

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.android.fhir.p2p
+
+import ca.uhn.fhir.context.FhirContext
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.descriptors.PrimitiveKind
+import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+import org.hl7.fhir.instance.model.api.IBaseResource
+
+/** The payload for syncing changes between peers. */
+@Serializable
+data class SyncPayload(
+  /** The current version of the resource after all of the changes have been applied. */
+  @Serializable(with = FhirResourceSerializer::class) val resource: IBaseResource,
```

Comment:
Is `IBaseResource` the level that we want to hook into the hapi model. `engine` uses [Resource](https://github.com/hapifhir/org.hl7.fhir.core/blob/master/org.hl7.fhir.r4/src/main/java/org/hl7/fhir/r4/model/Resource.java) throughout?

Reply:
I don't think that we can go any higher because `IParser.parseResource` returns `IBaseResource` unless we specify the resource type which we don't know ahead of time.